### PR TITLE
{2023.06}[foss/2022a] DP3 6.0 and WSClean 3.4

### PR DESCRIPTION
**Pull request by t1mk1k**
Part of this request: https://gitlab.com/eessi/support/-/issues/14#note_1610960931

### CODE REVIEW

**Comment by casparvl** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell

**Comment by casparvl** [comment]
bot: show_config

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/intel/haswell

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by casparvl** [comment]
Fails while installing `wget`:

```
checking for OPENSSL... no
configure: error: Package requirements (openssl) were not met:

No package 'openssl' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables OPENSSL_CFLAGS
and OPENSSL_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
 (at easybuild/tools/run.py:681 in parse_cmd_output)
```
I guess our EESSI container is _really_ bare... :D

**Comment by casparvl** [comment]
Ah, I see
```
# make sure pkg-config picks up system packages (OpenSSL & co)
preconfigopts = "export PKG_CONFIG_PATH=/usr/lib64/pkgconfig:/usr/lib/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig && "
configopts = '--with-ssl=openssl '
```
in the `wget` EasyConfig. That is probably not compatible with using `sysroot` :D 

I guess we can do two things:

1. Change this in a hook, specifically for EESSI
2. Create an EasyBlock for `wget` that sets `preconfigopts`, and that takes `sysroot` into account

Option 2 is clearly the more generic one. But, I'll start by modifying the EasyConfig itself, and seeing if it installs then, just to be sure that this is causing the issue.

**Comment by casparvl** [comment]
Adapting the `EasyConfig` to

```
# make sure pkg-config picks up system packages (OpenSSL & co)
preconfigopts = "export PKG_CONFIG_PATH=/cvmfs/pilot.eessi-hpc.org/versions/2023.06/compat/linux/x86_64/usr/lib64/pkgconfig && "
```

works, so this was indeed the problem, and both option 1 and 2 should provide working solutions.

**Comment by boegel** [comment]
The problem with `wget` is fixed in https://github.com/easybuilders/easybuild-easyconfigs/pull/19080, but that requires a small enhancement in EasyBuild framework, see https://github.com/easybuilders/easybuild-framework/pull/4359, which will be included in the upcoming EasyBuild v4.8.2 release which I hope to get out in the next couple of days.

Once EasyBuild v4.8.2 is out, we should:
* i) install EasyBuild v4.8.2 in the EESSI 2023.06 pilot repo (similar to #333);
* ii) create a new YAML file for `2022a` easyconfigs installed with EasyBuild v4.8.2 for EESSI 2023.06 (`eessi-2023.06-eb-4.8.1-2022a.yml`);

Since the changes in https://github.com/easybuilders/easybuild-easyconfigs/pull/19080 will also be included in EasyBuild v4.8.2, there's no need to use `--from-pr` for `wget`.

**Comment by boegel** [comment]
Since the easyconfigs contributed in will also be included in EasyBuild v4.8.2, the contents of `eessi-2023.06-eb-4.8.1-2022a.yml` can just be:

```yaml
easyconfigs:
  - DP3-6.0-foss-2022a:
  - WSClean-3.4-foss-2022a:
```

**Comment by t1mk1k** [comment]
@boegel and @casparvl, I have added an EasyStack file for EasyBuild 4.8.2. I'm not sure if I did it correctly, as the code seems to have changed from the 4.8.1 + Java/11 PR.

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by casparvl** [comment]
Hm....
```
== ... (took < 1 sec)
  >> running command:
        [started at: 2023-10-30 10:37:44]
        [working dir: /project/60006/SHARED/jobs/2023.10/pr_370/event_b5928530-770e-11ee-98dd-2642c8fc92f1/run_000/linux_x86_64_amd_zen2/eessi-hpc.org-2023.06-software]
        [output logged in /tmp/eb-l4nzqpvd/easybuild-run_cmd-bn5pjy5o.log]
        bzip2 /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/EasyBuild/4.8.2/easybuild/easybuild-EasyBuild-4.8.2-20231030.103744.log
  >> command completed: exit 0, ran in 00h00m09s
== COMPLETED: Installation ended successfully (took 1 min 47 secs)
== Results of the build can be found in the log file(s) /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/EasyBuild/4.8.2/easybuild/easybuild-EasyBuild-4.8.2
-20231030.103744.log.bz2
== Build succeeded for 1 out of 1
== Temporary log file(s) /tmp/eb-l4nzqpvd/easybuild-kzz3q8n3.log* have been removed.
== Temporary directory /tmp/eb-l4nzqpvd has been removed.
ESC[32mLatest EasyBuild release installed, let's go!ESC[0m
ESC[33m>> Still no module for EasyBuild v4.8.2, trying with easyconfig EasyBuild-4.8.2.eb...ESC[0m
ESC[31mERROR: No easyconfig found for EasyBuild v4.8.2ESC[0m
Leaving Gentoo Prefix with exit status 1
```
Having a new EasyBuild deployed _and_ using it immediately seems to have worked in https://github.com/EESSI/software-layer/pull/333 , but here it seesm to not find the module it just installed. It fails to find it in [this line](https://github.com/EESSI/software-layer/blob/3b55b9e9bbc56ef801ba6ef87d6c0fd1f5355f8c/load_easybuild_module.sh#L74)

I guess it might be caching, we could alter that script and insert an

```
module avail 2>&1 | grep -i easybuild/${EB_VERSION} &> ${ml_av_easybuild_out}
if [[ $? -eq 0 ]]; then
    echo_green ">> Module for EasyBuild v${EB_VERSION} found!"
else
    module avail --ignore_cache 2>&1 | grep -i easybuild/${EB_VERSION} &> ${ml_av_easybuild_out}
    if [[ $? -eq 0 ]]; then
        echo_green ">> Module for EasyBuild v${EB_VERSION} found!"
    else
       <the original 'else' block here>
   fi
fi
...
```
That way, we first try the cached `module avail` (which is fastest, and should give the EB module _unless_ the PR added an EB version). If that fails, we try uncached. If that feels, we try to install (i.e. the original 'else' block).

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by casparvl** [comment]
Same error, frustrating...
```
== COMPLETED: Installation ended successfully (took 1 min 39 secs)
== Results of the build can be found in the log file(s) /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/EasyBuild/4.8.2/easybuild/easybuild-EasyBuild-4.8.2
-20231030.122703.log.bz2
== Build succeeded for 1 out of 1
== Temporary log file(s) /tmp/eb-yua_2ikb/easybuild-vl_2j3lb.log* have been removed.
== Temporary directory /tmp/eb-yua_2ikb has been removed.
ESC[32mLatest EasyBuild release installed, let's go!ESC[0m
ESC[33m>> Still no module for EasyBuild v4.8.2, trying with easyconfig EasyBuild-4.8.2.eb...ESC[0m
ESC[31mERROR: No easyconfig found for EasyBuild v4.8.2ESC[0m
Leaving Gentoo Prefix with exit status 1
```
I don't understand how it is not found. The module is clearly there:
```
== creating module...
  >> generating module file @ /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/modules/all/EasyBuild/4.8.2.lua
== ... (took 1 secs)
```
and that path should already be on the `MODULEPATH` (it's where `EasyBuild 4.8.1` that was used to install this was also already located)

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

File: load_easybuild_module.sh [hunk 1]
```diff
@@ -75,17 +75,22 @@ else
...
-        echo_yellow ">> Still no module for EasyBuild v${EB_VERSION}, trying with easyconfig ${eb_ec}..."
-        ${EB} --search ${eb_ec} | grep ${eb_ec} > /dev/null
+        module --ignore_cache avail 2>&1 | grep -i easybuild/${EB_VERSION} &> ${ml_av_easybuild_out}
         if [[ $? -eq 0 ]]; then
-            echo "Easyconfig ${eb_ec} found for EasyBuild v${EB_VERSION}, so installing it..."
-            ok_msg="EasyBuild v${EB_VERSION} installed, alright!"
-            fail_msg="Installing EasyBuild v${EB_VERSION}, yikes! (output: ${eb_install_out})"
-            ${EB} EasyBuild-${EB_VERSION}.eb 2>&1 | tee -a ${eb_install_out}
-            check_exit_code $? "${ok_msg}" "${fail_msg}"
+            echo_green ">> Module for EasyBuild v${EB_VERSION} found!"
```
Comment (casparvl) [inline load_easybuild_module.sh:80]: You'll need to add another line:

IGNORE_CACHE='--ignore_cache'  # Module load should also ignore the Lmod cache

And then below (line 111 I believe:

```
module ${IGNORE_CACHE} load EasyBuild/${EB_VERSION}
```

File: load_easybuild_module.sh [hunk 2]
```diff
@@ -75,17 +75,22 @@ else
     if [[ $? -eq 0 ]]; then
         echo_green ">> Module for EasyBuild v${EB_VERSION} found!"
```
Comment (casparvl) [inline load_easybuild_module.sh:76]: Probably, we should also set `IGNORE_CACHE=''` here, just to be sure it is an empty string (since we will rely on that in line `111`, when we do the `module ${IGNORE_CACHE} load ...`.

**Review by casparvl — CHANGES_REQUESTED** [review]

File: load_easybuild_module.sh (hunk 2 above)
Comment (casparvl) [reply load_easybuild_module.sh:76]: Better yet, do it on line `74` so that `''` is the default value for all cases. Then, the second case (where we actually ignore the cache to find the module) is the only exception - and that then overwrites `IGNORE_CACHE='--ignore_cache'`

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

File: load_easybuild_module.sh (hunk 1 above)
Comment (t1mk1k) [reply load_easybuild_module.sh:81]: I added the line too early, so moved it down.

**Comment by casparvl** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by laraPPr** [comment]
@casparvl @t1mk1k should we maybe split this pr up and open a new one for Easybuild v4.8.2. Because we also need that framework fix for Pillow in these two prs as well.
* [#362](https://github.com/EESSI/software-layer/pull/362) 
* [#339](https://github.com/EESSI/software-layer/pull/339)

**Comment by casparvl** [comment]
@laraPPr we could, but I really wanted to resolve the issue with the cache as well - otherwise, someone might run into it next time again :) It seems that was resolved now: I'm currently looking at the build logs for job `390` and it's happily building the stuff in `eessi-2023.06-eb-4.8.2-2022a.yml`. Let's see how that goes. If it completes for this architecture, I'll try to start it tonight for all the others as well. If that all completes succesfully, there's no need to do a seperate build: we can just deploy this one tomorrow. I'm guessing those two other PRs can wait for one more night :)

If it _doesn't_ complete successfully for whatever reason, we'll split off EasyBuild as a seperate PR (including the fix for ignoring the cache, so that we don't run into that in the future) tomorrow morning. We can then get that merged asap.

**Comment by casparvl** [comment]
Hm, ok, that's our cue: time to split of the EasyBuild 4.8.2 PR...

@t1mk1k : the error:

```
CMake Warning at measures/CMakeLists.txt:8 (message):
  No ephemerides data found in the specified DATA_DIR
  ("/cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/casacore/3.5.0-foss-2022a/data"),
  so the location must be specified at runtime.  If you have preinstalled
  casacore data, specify the location with -DDATA_DIR.


-- Looking for python3 specific environment...
CMake Error at /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/CMake/3.24.3-GCCcore-11.3.0/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  version "3.12.0")
Call Stack (most recent call first):
  /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/CMake/3.24.3-GCCcore-11.3.0/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/CMake/3.24.3-GCCcore-11.3.0/share/cmake-3.24/Modules/FindPython/Support.cmake:3203 (find_package_handle_standard_args)
  /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/amd/zen2/software/CMake/3.24.3-GCCcore-11.3.0/share/cmake-3.24/Modules/FindPython3.cmake:490 (include)
  python3/CMakeLists-cmake3.14.txt:3 (find_package)
  python3/CMakeLists.txt:4 (include)


-- Configuring incomplete, errors occurred!
See also "/tmp/bot/easybuild/build/casacore/3.5.0/foss-2022a/easybuild_obj/CMakeFiles/CMakeOutput.log".
See also "/tmp/bot/easybuild/build/casacore/3.5.0/foss-2022a/easybuild_obj/CMakeFiles/CMakeError.log".
```

I think we've had some earlier isuses finding `python3` as well. I need to find out where that was though, might be a PR that is still open...

**Comment by casparvl** [comment]
Looking at https://github.com/easybuilders/easybuild-easyconfigs/pull/18963/files we specified the python interpreter directly on the `configopts`. We did this by modifying the EasyConfigs. 

The quickest way to see _if_ this helps, is probably to first interactively reproduce the problem [according to these instructions](https://github.com/EESSI/docs/pull/117) (note: work in progress, but should work already :)), then modifying your EasyConfig locally (by adding an additional `-DPYTHON_SOMETHING_SOMETHING=$EBROOTPYTHON/bin/python` to the `configopts`) and see if that makes it pass. If it does, you can update the EasyConfig in a PR to EasyBuild (as this extra configopt shouldn't hurt _anyone_, in fact, it might prevent some incorrect python's being picked up for others too).

**Comment by casparvl** [comment]
Btw, not sure if that `Warning` at the top was expected? You probably know better :) maybe something went wrong there as well? It wouldn't block the build since it is a warning, but might mean something will be missing?

**Comment by t1mk1k** [comment]
@casparvl and @laraPPr I've created a separate PR for EasyBuild v4.8.2 with the cache fix: https://github.com/EESSI/software-layer/pull/376

**Comment by casparvl** [comment]
@t1mk1k  Can you sync with `2023.06` branch upstream? (you might need to discard your changes that include the EasyBuild 4.8.2 and the `load_easybuild_modules.sh`)

**Comment by t1mk1k** [comment]
While debugging using the EESSI container, I have run into two issues which I'm still in the process of solving.

First, configuring AOFlagger results in the following error:

```
CMake Error at /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/generic/software/CMake/3.24.3-GCCcore-11.3.0/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR) (Required is at
  least version "5.3")
```

This is because Lua is filtered out of the dependency list because of the EasyBuild configuration in the EESSI container:

```
[EESSI pilot 2023.06] $ eb --show-config
#
# Current EasyBuild configuration
# (C: command line argument, D: default value, E: environment variable, F: configuration file)
#

...

filter-deps          (E) = Autoconf, Automake, Autotools, binutils, bzip2, DBus, flex, gettext, gperf, help2man, intltool, libreadline, libtool, Lua, M4, makeinfo, ncurses, util-linux, XZ, zlib

...
```

I will create a hook to add an additional configopts to point to the Lua version in the compat layer, as this is an EESSI specific issue.

The second issue is that several packages seem to pick up the (newer) Python version in the compat layer, instead of the Python version listed in the dependencies. 
This seems to be caused by CMake's FindPython modules, which first [load the newest Python version](https://github.com/Kitware/CMake/blob/master/Modules/FindPython.cmake#L320) (regardless of PATH priority) and [stop as soon as the version requirement is satisfied](https://github.com/Kitware/CMake/blob/d1906daebfaf1938ae9c6abc4e2288e7a3f581cf/Modules/FindPython.cmake#L324C1-L326C71).
In other words, if no version requirements (or find strategy) are specified as arguments to the FindPython function by a developer, there is a risk that the system's version of Python will be selected instead of the version listed in the dependencies, when the systems' Python version is newer.

**Comment by t1mk1k** [comment]
@casparvl Unfortunately, the Lua version in the compat layer is 5.1 but the software requires at least Lua 5.3, so from my end I'm unable to fix this issue.
In addition, I saw that `ncurses` is also filtered from the dependency list, whereas this is one of casacore's dependencies.

File: eessi-2023.06-eb-4.8.2-2022a.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+easyconfigs:
+  - AOFlagger-3.4.0-foss-2022a:
+      options:
+        from-pr: 19119
+        # Exclude Lua
```
Comment (ocaisa) [inline eessi-2023.06-eb-4.8.2-2022a.yml:5]:
```suggestion
        # Exclude Lua from `filter-deps` as the compat layer version is too old for the software
```

**Comment by ocaisa** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by ocaisa** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2

**Comment by ocaisa** [comment]
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/generic
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi-hpc.org-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi-hpc.org-2023.06-software arch:aarch64/generic
bot: build repo:eessi-hpc.org-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi-hpc.org-2023.06-software arch:aarch64/neoverse_v1

**Review by ocaisa — APPROVED** [review]

**Comment by ocaisa** [comment]
@t1mk1k This is now ingested and available, thanks for your patience!
